application/common/models/Method.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace common\models;
+
+use common\exceptions\InvalidCodeException;
+use common\helpers\MySqlDateTime;
+use common\helpers\Utils;
+use yii\helpers\ArrayHelper;
+use yii\web\ConflictHttpException;
+use yii\web\ServerErrorHttpException;
+
+/**
+ * Class Method
+ * @package common\models
+ *
+ */
+class Method extends MethodBase
+{
+    public function rules()
+    {
+        return ArrayHelper::merge(
+            [
+                [
+                    ['uid'], 'default', 'value' => Utils::generateRandomString(),
+                ],
+
+                [
+                    ['verified', 'verification_attempts'], 'default', 'value' => 0,
+                ],
+
+                [
+                    'verification_code', 'default', 'when' => function() { return $this->getIsNewRecord(); },
+                    'value' => Utils::getRandomDigits(\Yii::$app->params['reset']['codeLength']),
+                ],
+
+                [
+                    'verification_expires', 'default', 'when' => function() { return $this->getIsNewRecord(); },
+                    'value' => MySqlDateTime::formatDateTime(time() + \Yii::$app->params['reset']['lifetimeSeconds']),
+                ],
+
+                [
+                    'value', 'email',
+                ],
+
+                [
+                    ['created'], 'default', 'value' => MySqlDateTime::now(),
+                ],
+
+            ],
+            parent::rules()
+        );
+    }
+
+    public function fields()
+    {
+        return [
+            'id' => function() { return $this->uid; },
+            'value',
+        ];
+    }
+
+    /**
+     * @return string
+     * @throws \Exception
+     */
+    public function getMaskedValue()
+    {
+        return Utils::maskEmail($this->value);
+    }
+
+    public function isVerified()
+    {
+        return $this->verified === 1 ? true : false;
+    }
+
+    /**
+     * Send verification message
+     * @throws ConflictHttpException
+     * @throws ServerErrorHttpException
+     */
+    public function sendVerification()
+    {
+        if ($this->isVerified()) {
+            throw new ConflictHttpException('Method already verified', 1540689804);
+        }
+
+        $this->verification_attempts++;
```

Comment:
Sorry to be rehashing existing design, by why does sending a verification email update the `verification_attempts` counter? Isn't that supposed to be the number of times someone has submitted a value in an attempt to verify the Method?

Reply:
I think I agree. Only thing I can think of is the meaning given to `maxAttempts` is defined this way. Who sets the instance-specific configuration, and what information do they use to understand the meaning of the variables? I will go ahead and remove this line and change the test accordingly, but if you or @fillup have any additional background information, I'll change it back.

Reply:
We may have it this way so if someone kept switching between methods without attempting to verify it we'd at least stop it from becoming a spam bot. We can have a reasonably high limit to ensure the ability to switch methods and click links, but why not prevent them from switching 200 times and generating a lot of emails?

Reply:
I will leave the code as it was, then, so that the attempts counter will increment when the email is sent as well as when verification attempts are made.